rfcs/0032-blockchain-in-hard-fork

```diff
@@ -0,0 +1,153 @@
+## Summary
+[summary]: #summary
+
+We present alternatives for handling the blockchain in a hard fork. The focus is
+on hard forks in response to a network failure, though the mechanisms may be
+applicable in more general scenarios.
+
+This RFC does not consider how the transaction SNARK and scan state are handled
+in hard forks. There will be another RFC for those, and an RFC to describe
+the entire hard fork plan.
+
+## Motivation
+[motivation]: #motivation
+
+For the launch of the Coda main net, we wish to have a hard fork plan and implementation
+in place, in case there's a network failure. One of the parts of such a plan
+is how to update the blockchain.
+
+## Summary description of alternatives
+[summary-descriptions]: #summary-descriptions
+
+After the fork, do one of these:
+
+1. New genesis timestamp and genesis ledger; assert a new starting epoch and slot
+
+2. Retain the original genesis timestamp, with no time offset, with rules to
+    maintain chain strength
+
+3. Retain the original genesis timestamp, use a time offset for new blocks
+
+## Detailed description of alternatives
+[detailed-descriptions]: #detailed-descriptions
+
+Alternative 1
+=============
+
+- choose a root and ledger at the pause point
+
+- the initial epoch, slot are the special case of 0,0
+   - otherwise, increment the slot from the pause point
+   - provide as a compile-time parameter
+
+- if we have a proof of the protocol state, provide that
+
+- subsequent blocks may use different proof, verification keys
+
+- force a software upgrade by bumping the protocol major version
+
+Notes
+-----
+
+- nodes will bootstrap from the new genesis state
+
+- time-locked accounts and transactions with expiration should continue
+   to work, because they rely on global slot numbers, and the
+   slot is incremented from the pause point
+
+Issues
+------
+
+- what is the source of randomness (epoch seed) for the post-fork chain?
+
+- is it acceptable to erase history (but we already have a succinct blockchain)
+
+- specifically, is there unfairness to the chains discarded past the pause point
+
+- what is allowed to change after the fork, with what effect:
+   - protocol state structure
+   - ledger structure/depth,
+   - account structure
+   - consensus constants
+
+- do we have to maintain pre-, post-fork code/data structures to verify the new genesis state,
+   ledger, and new blocks
+
+- if there's no proof of the ledger, how to convince the world that the new genesis ledger and
+   ledger are valid
+
+Alternative 2
+=============
+
+- choose a root and ledger at a pause point
+
+- nodes have a notion of "last fork time", an epoch-and-slot
+
+- upon a fork, gossip a special block to update the last fork time is issued, also referring to the
```

Comment:
Does this proposal work in the world where the "update protocol state" function changes considerably? If so, it is not clear to my to what extent this differs from option 1.

Reply:
Is there a general rule for when to gossip a special block vs. when to force upgrade the client?

Reply:
If the structure of the protocol state or ledger changes, that would require updating the software.

Reply:
We should always gossip a fork block when we fork. The client will need a force upgrade if the protocol has changed as part of that fork, otherwise not.

Reply:
Thanks. Also to clarify, for Alternatives 2 and 3, what will the special block include? I was thinking it should just be two things: failure time and fork time. Were y'all thinking differently? @psteckler @mrmr1993

Reply:
To make the transition transparent, we will probably want:
* the previous protocol state + verification key (ie. proof inputs) for the last pre-fork block
  - implicitly this contains the failure time
* whether the fork was safe/unsafe
* the resume time
* the new protocol state + verification key
  - (actually, all of the above will be encoded in the fork state within the protocol state anyway, so that they are accessible to the proof)

Reply:
Thanks. It's not clear to me why we need to include the protocol state and the verification key. There is likely no consensus on the latest protocol state and verification key among all honest parties. If possible, we should try to avoid forcing anything that hasn't reached a consensus yet.   

I was instead thinking that this special block simply contains the failure time and the restart time. This is the only thing we know for sure and can force upon others. Nodes can build off of it as follows: They take their protocol state + verification key (ie. proof inputs) for the last  block before the failure time stated in the special block and the first block after the restart time treats the former block as its previous block. 

Note that I am proposing that the block we start off from after the restart is not necessarily the last pre-fork block, as there is a chance that some part of the network hasnt failed and is continuing to generate blocks after the "failure time". For consistency, we just ask everyone to subscribe to the same failure time and fork time. This is not unfair, because, the blocks generated post-failure and pre-restart will face the same "view" so to speak as if there were no fork, because the block winning probabilities are determined by the same values.

Reply:
> It's not clear to me why we need to include the protocol state and the verification key. There is likely no consensus on the latest protocol state and verification key among all honest parties.

Perhaps I have confused my terminology: I've been thinking in terms of the proof(s) that the network can use as a starting point to recover after a disaster, where we want to expose these to signal that this is (in some sense) a continuation of a pre-fork chain.

> Note that I am proposing that the block we start off from after the restart is not necessarily the last pre-fork block, as there is a chance that some part of the network hasnt failed and is continuing to generate blocks after the "failure time".

I had imagined that the disaster recovery scenario meant that the network couldn't make meaningful progresses along any chain for some reason; in that situation, I struggle to see how emitting a 'license to skip' message would improve things. Is there another disaster scenario that we should be considering instead/too?

Reply:
I think I got the terminology correct, I am just saying the following. Let B_last be the last block you had before the failure. The last block I had at that time may not be the same. (In fact, the one I have may not even make it to the blockchain eventually. The only blocks we are sure all honest folks have are the ones that are k or more deep in history).

Reply:
Regarding disaster scenarios, good point. Consider the following scenario. Between June 16 and June 20, there is an internet outage with 95% of the Coda clients. So these folks are not able to "advance their chain". The other 5% is generating blocks during those days. The internet is back on track everywhere on June 21st. We circulate a special block to everyone on June 21 that "there was an outage between June 16 and June 20, so please squash that time period to 0". Why would we like to do this? This is so that (1) those who had outage know that "they didnt miss anything" (2) those who didnt have the outage should suffer the loss of creating honest blocks during the time others had the outage; the idea of pausing the absolute time as far as the blockchain is concerned ensures that they get the same opportunity after the restart to create those blocks.

Reply:
Let me know if you agree with the following assumption: A failure can be because of one of the two things: network or protocol. When it's the network, there is very little you can do during the outage; just when it comes back up, we gossip the special block. When it's the protocol, we force an upgrade... So when it's a network failure, we dont need to gossip anything beyond that there was a network failure. Everyone is supposed to have every other information.